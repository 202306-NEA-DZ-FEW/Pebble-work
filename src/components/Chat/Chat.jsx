import {
    arrayUnion,
    collection,
    doc,
    getDoc,
    onSnapshot,
    orderBy,
    query,
    updateDoc,
} from "firebase/firestore";
import { useEffect, useState } from "react";

import { auth, db } from "@/util/firebase";

const Chat = () => {
    const [messages, setMessages] = useState([]);
    const [newMessage, setNewMessage] = useState("");
    const user = auth.currentUser;

    // Listen for new messages
    useEffect(() => {
        const q = query(collection(db, "messages"), orderBy("timestamp"));

        const unsubscribe = onSnapshot(q, (snapshot) => {
            setMessages(snapshot.docs.map((doc) => doc.data()));
        });

        return unsubscribe;
    }, []);

    // Send a new message
    // Send a new message
    // Send a new message
    // Send a new message
    const sendMessage = async (e) => {
        e.preventDefault();

        if (user) {
            const chatRef = doc(db, "database", "Chat");

            const newMessageData = {
                text: newMessage,
                timestamp: new Date().getTime(),
                uid: user.uid,
            };

            // Add the new message to the database
            await updateDoc(chatRef, {
                messages: arrayUnion(newMessageData),
            });

            setNewMessage("");

            // Get the current chat data
            const chatDoc = await getDoc(chatRef);
            const chatData = chatDoc.data();

            // Filter out messages that are older than 40 seconds
            const validMessages = chatData.messages.filter((message) => {
                const messageAge = newMessageData.timestamp - message.timestamp;
                return messageAge <= 40000;
            });

            // Update the database with the valid messages
            await updateDoc(chatRef, { messages: validMessages });
        } else {
            console.log("you are not signed in");
        }
    };

    return (
        <>
            <div
                className='min-w-full'
                style={{ position: "relative", display: "inline-block" }}
            >
                <form onSubmit={sendMessage}>
                    <textarea
                        className='min-w-full resize-none scrollBar'
                        value={newMessage}
                        onChange={(e) => setNewMessage(e.target.value)}
                        onKeyDown={(e) => {
                            if (e.key === "Enter" && !e.shiftKey) {
                                e.preventDefault();
                                sendMessage(e);
                            }
                        }}
                        style={{
                            paddingRight: "30px",
                            height: "80px",
                        }}
                    ></textarea>
                    <button
                        className='w-6'
                        type='submit'
                        style={{
                            position: "absolute",
                            right: "4px",
                            top: "4px",
                        }}
                    >
                        <img src='/icons/message.svg' alt='Send' />
                    </button>
                </form>
            </div>
        </>
    );
};

export default Chat;
