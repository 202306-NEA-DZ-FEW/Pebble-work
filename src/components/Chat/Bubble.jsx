import React, { useState } from "react";
import styles from "@/styles/Bubble.module.css";
import ChatDisplay from "./ChatDisplay";
import { IoChatbubblesOutline } from "react-icons/io5";

const Bubble = () => {
    const [expanded, setExpanded] = useState(false);
    const [newMessage, setNewMessage] = useState(false);

    const handleButtonClick = () => {
        setExpanded(!expanded);
        if (!expanded) {
            setNewMessage(false); // reset new message state when chat is opened
        }
    };
    const handleNewMessage = () => {
        if (!expanded) {
            setNewMessage(true); // only set new message state if chat is not opened
        }
    };

    return (
        <div>
            <button
                className={`${styles.expandingButton} ${
                    newMessage ? styles.newMessage : ""
                }`}
                onClick={handleButtonClick}
            >
                {expanded ? "Close Chat" : "Open Chat"}
            </button>
            <div
                className={`flex items-end ${styles.expandingDiv} ${
                    expanded ? styles.expanded : ""
                }`}
            >
                <ChatDisplay onNewMessage={handleNewMessage} />
            </div>
        </div>
    );
};

export default Bubble;
