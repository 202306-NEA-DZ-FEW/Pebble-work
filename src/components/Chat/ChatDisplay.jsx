import { doc, getDoc, onSnapshot } from "firebase/firestore";
import { useEffect, useState } from "react";

import Chat from "./Chat";
import { auth, db } from "../../util/firebase";

const ChatDisplay = ({ onNewMessage }) => {
    const [messages, setMessages] = useState([]);
    const [currentTime, setCurrentTime] = useState(new Date().getTime());
    const getUserName = async (uid) => {
        const userRef = doc(db, "users", uid);
        const userDoc = await getDoc(userRef);

        return userDoc.data().Name;
    };

    // Listen for new messages
    useEffect(() => {
        const chatRef = doc(db, "database", "Chat");

        const unsubscribe = onSnapshot(chatRef, async (docSnapshot) => {
            const chatData = docSnapshot.data();

            const validMessages = chatData.messages.filter((message) => {
                // Check if the message was sent more than 180 seconds ago
                const messageAge = currentTime - message.timestamp;
                return messageAge <= 180000;
            });

            for (let message of validMessages) {
                message.userName = await getUserName(message.uid);
            }

            // Check if there are new messages
            if (validMessages.length > messages.length) {
                onNewMessage(); // call the callback when a new message arrives
            }

            setMessages(validMessages);
        });

        return () => unsubscribe();
    }, [currentTime, messages.length]);

    // Update the current time every second
    useEffect(() => {
        const timer = setInterval(() => {
            setCurrentTime(new Date().getTime());
        }, 1000);

        // Clean up the interval when the component unmounts
        return () => clearInterval(timer);
    }, []);

    return (
        <>
            <div className='flex relative pl-2 justify-center max-h-full w-full scrollBar overflow-y-scroll'>
                <div className='w-full'>
                    {messages.map((message, i) => {
                        // Calculate the age of the message in seconds
                        const messageAge =
                            (currentTime - message.timestamp) / 1000;

                        // Calculate the opacity based on the age of the message
                        let style = {};
                        if (messageAge > 178) {
                            const transitionDuration = 181 - messageAge;
                            style = {
                                opacity: 0,
                                transition: `opacity ${transitionDuration}s linear`,
                            };
                        }

                        // Determine the alignment of the message based on the user
                        const alignment =
                            message.uid === auth.currentUser.uid
                                ? "chat chat-end"
                                : "chat chat-start";

                        return (
                            <div
                                key={i}
                                className={`overflow-hidden ${alignment}`}
                            >
                                <div
                                    className='chat-bubble overflow-x-hidden break-words'
                                    style={style}
                                >
                                    <strong>{message.userName}</strong>:{" "}
                                    {message.text}
                                </div>
                            </div>
                        );
                    })}
                    <Chat />
                </div>
            </div>
        </>
    );
};

export default ChatDisplay;
