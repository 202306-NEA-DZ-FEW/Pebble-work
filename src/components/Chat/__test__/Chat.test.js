import React from "react";
import renderer from "react-test-renderer";

import Chat from "../Chat";

describe("Chat component", () => {
    it("renders correctly", () => {
        const tree = renderer.create(<Chat />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
