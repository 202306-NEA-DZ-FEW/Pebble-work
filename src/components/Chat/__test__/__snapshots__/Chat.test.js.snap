// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat component renders correctly 1`] = `
<div
  className="min-w-full"
  style={
    {
      "display": "inline-block",
      "position": "relative",
    }
  }
>
  <form
    onSubmit={[Function]}
  >
    <textarea
      className="min-w-full resize-none scrollBar"
      onChange={[Function]}
      onKeyDown={[Function]}
      style={
        {
          "height": "80px",
          "paddingRight": "30px",
        }
      }
      value=""
    />
    <button
      className="w-6"
      style={
        {
          "position": "absolute",
          "right": "4px",
          "top": "4px",
        }
      }
      type="submit"
    >
      <img
        alt="Send"
        src="/icons/message.svg"
      />
    </button>
  </form>
</div>
`;
